### PR TITLE
Prometheus metrics

### PR DESCRIPTION
**Pull request by ath88**

### CODE REVIEW

**Comment by ath88** [comment]
dont review this. express-prometheus-middleware is deprecated

**Comment by ath88** [comment]
https://github.com/joao-fontenele/express-prometheus-middleware/issues/71

**Comment by ath88** [comment]
Hang on while this hopefully resolves quickly.

https://github.com/jochen-schweizer/express-prom-bundle/pull/101

**Comment by ath88** [comment]
@hypesystem Ready for review

File: package.json [hunk 1]
```diff
@@ -4,12 +4,13 @@
     "dexpress-finalhandler": "^0.1.2",
     "express": "^4.18.1",
     "express-async-handler": "^1.2.0",
-    "express-prometheus-middleware": "^1.2.0",
+    "express-prom-bundle": "^6.5.0",
```
Comment (hypesystem) [inline package.json:7]: I see only express-prom-bundle, but also a dexpress-prom-bundle in the package-lock file? We might need a lockfile cleanup

File: index.js [hunk 2]
```diff
@@ -1,12 +1,12 @@
 import express from 'express';
 import setUpApplicationDefaults from './setUpApplicationDefaults.js';
 
-async function proxy() {
-    const app = express.apply(express, arguments);
+async function proxy(config = { }) {
```
Comment (hypesystem) [inline index.js:4]: Syntax: prefer `{}` over `{ }` here :smile:

File: setUpApplicationDefaults.js [hunk 3]
```diff
@@ -61,7 +61,25 @@ export default async (app) => {
...
+    if (config.prometheusMetrics) {
+        const metricsConfig = { ...config.prometheusMetrics };
+        if (config.prometheusMetrics.port) {
+            app.metricsApp = express();
+            app.metricsApp.listen(metricsConfig.port);
+            metricsConfig.metricsApp = app.metricsApp;
+            delete metricsConfig.port;
+        }
+        app.use(promBundle({
+          metricsApp: app.metricsApp,
```
Comment (hypesystem) [inline setUpApplicationDefaults.js]: What is metricsApp in this case, if `port` is falsy?

Does it make sens to listen to the app right off the bat, too? I think that seems a little over-eager to me. I would prefer changing the main `app.listen` to first start the metricsApp listening instead? That would align closer to my expected behavior.

File: setUpApplicationDefaults.js [hunk 4]
```diff
@@ -61,7 +61,25 @@ export default async (app) => {
...
+          autoregister: false,
+          includeMethod: true,
+          includePath: true,
+          promClient: { collectDefaultMetrics: { } }
```
Comment (hypesystem) [inline setUpApplicationDefaults.js]: Now that we're talking about changes, I think supporting more options in `createApplication` would be nice - maybe we can expose these options somehow?

File: setUpApplicationDefaults.js [hunk 5]
```diff
@@ -61,7 +61,25 @@ export default async (app) => {
...
+        }));
+    }
+    // TODO we need to terminate app.metricsApp too somehow on shutdown
```
Comment (hypesystem) [inline setUpApplicationDefaults.js]: Just like I suggested with overwriting app.listen, I think we should do the same for server.close, probably...

**Review by hypesystem — CHANGES_REQUESTED** [review]
Some significant questions about behavior vs expected behavior.

I generally love the approach, I just want to make sure we hit a good/easy to use solution here.

File: setUpApplicationDefaults.js (hunk 5 above)
Comment (ath88) [reply setUpApplicationDefaults.js]: I see a problem in trying to extend server.close. It's a deep native thing, so making it magic feels weird to me.

I don't really like patching native code with too much magic, so we need to take a stand here. Same problem with app.listen, which is not a promise. We could make it a promise, but we would be breaking expectations here.

File: package.json (hunk 1 above)
Comment (ath88) [reply package.json:7]: You're right. [326c423](https://github.com/derangeddk/dexpress/pull/4/commits/326c42369f132805271a57a9e8e89675eb371791)

File: index.js (hunk 2 above)
Comment (ath88) [reply index.js:4]: I prefer `{ }`. :D We could put this into a linter-rule so we have a forum for more general discussions and agreements on this. See https://gitlab.deranged.dk/deranged/eslint-plugin

File: setUpApplicationDefaults.js (hunk 3 above)
Comment (ath88) [reply setUpApplicationDefaults.js]: It's undefined, which is fine for promBundle.

I've changed startup to happen during `app.listen` instead. [a059624](https://github.com/derangeddk/dexpress/pull/4/commits/a059624936079571ab93999a11d532cbf6f99c21)

File: setUpApplicationDefaults.js (hunk 4 above)
Comment (ath88) [reply setUpApplicationDefaults.js]: [9470cba](https://github.com/derangeddk/dexpress/pull/4/commits/9470cbae86b1fdc7cae4215d7d15b2223b067690)

File: index.js (hunk 2 above)
Comment (hypesystem) [reply index.js:4]: 😢

File: setUpApplicationDefaults.js (hunk 5 above)
Comment (hypesystem) [reply setUpApplicationDefaults.js]: Agreed -- but we need some way to ensure shutdown happens properly, right? What's the current solution to this?

Comment (hypesystem) [reply setUpApplicationDefaults.js:83]: Agreed -- but we need some way to ensure shutdown happens properly, right? What's the current solution to this?

Comment (ath88) [reply setUpApplicationDefaults.js]: [7885c55](https://github.com/derangeddk/dexpress/pull/4/commits/7885c55dea99045736851a9746fb817ec7c7a527)

**Comment by ath88** [comment]
@hypesystem Should be fixed now :)

File: setUpApplicationDefaults.js [hunk 6]
```diff
@@ -64,7 +69,25 @@ export default async (app) => {
...
+          autoregister: !Boolean(config.prometheusMetrics.port),  // if no port defined, we register on primary express app
+          includeMethod: true,
+          includePath: true,
+          promClient: {
+            collectDefaultMetrics: { },
+            ...config.prometheusMetrics.promClient, // see https://github.com/siimon/prom-client for config options
+          },
+        }));
+    }
+    // TODO we need to terminate app.metricsApp too somehow on shutdown
```
Comment (hypesystem) [inline setUpApplicationDefaults.js]: This is now handled, comment should be removed!

File: setUpApplicationDefaults.js [hunk 7]
```diff
@@ -64,7 +69,25 @@ export default async (app) => {
     // Attach always useful middleware
     app.use(helmet());
     app.use(cors());
-    app.use(prometheus());
+
+    // Add prometheus monitoring metrics
+    if (config.prometheusMetrics) {
```
Comment (hypesystem) [inline setUpApplicationDefaults.js]: Reasonable defaults: is it on purpose that we default to not having a prometheus metrics app at all? That may be reasonable enough, I haven't thought it through...

But I want the general approach of this lib to be "don't worry, we do the good thing per default, but you can config it or toggle it off" - I don't know if we're hitting that here with default-off.

**Review by hypesystem — APPROVED** [review]
Looks good! Generally approved, some few last questions, but not blocking for a merge and publish :smile:

File: setUpApplicationDefaults.js (hunk 6 above)
Comment (ath88) [reply setUpApplicationDefaults.js]: Done

File: setUpApplicationDefaults.js (hunk 7 above)
Comment (ath88) [reply setUpApplicationDefaults.js]: Agree. [5fc3eae](https://github.com/derangeddk/dexpress/pull/4/commits/5fc3eae4934ab82c51e608e42f021e368b014a15)

**Pull request merged**